fakeldap.py, tests.py: implemented anded conditional searching

ClundXIII opened the pull request:
Re-wrote `_simple_onelevel_search` into `_multiple_attrs_onelevel_search` which now takes a list of tuples aka a list of conditions that all have to be met in order for the element to be added to the result list.

Add a second regular expression to `_search_s` that allows multiple anded conditions to be processed: (&(attr1=val1)(attr2=val2)...)

Implement a short test for the new functionality.

diabolicvincent-git commented:
Very excited about this as it would allow LDAP sync for our community which is huge! Hoping for a speedy acceptance of this pull request :)

ClundXIII commented:
@mateuszmandera ping :)

mateuszmandera commented on fakeldap.py:
```diff
@@ -417,12 +417,21 @@ def _search_s(self, base, scope, filterstr, attrlist, attrsonly):
+
+            if r_simple.match(filterstr) is not None:
+                search_attr_name, search_attr_value = filterstr[1:-1].split('=')
+                return self._multiple_attrs_onelevel_search(base, [(search_attr_name, search_attr_value)])
+            elif r_anded.match(filterstr) is not None:
+                bracketList = filterstr[3:-2] # cut of the `(&(` and `))` --> attr1=val1)(attr2=val2
+                attrs_conditions = [condition.split("=") for condition in bracketList.split(")(")]
+                return self._multiple_attrs_onelevel_search(base, attrs_conditions)
```
Can we be consistent in  giving `_multiple_attrs_onelevel_search` a list of tuples?  `attrs_conditions = [condition.split("=") for condition in bracketList.split(")(")]` makes the arg a list of lists rather than tuples. The function's logic may not care but it'd be nice to have a consistent type

ClundXIII replied:
is this sufficiently?

409a02dc79ec5cadf92af1bd971dee12354c5a9d

```diff
diff --git a/fakeldap.py b/fakeldap.py
index f0967a9..03e8437 100644
--- a/fakeldap.py
+++ b/fakeldap.py
@@ -427,7 +427,7 @@ class MockLDAP(object):
                 return self._multiple_attrs_onelevel_search(base, [(search_attr_name, search_attr_value)])
             elif r_anded.match(filterstr) is not None:
                 bracketList = filterstr[3:-2] # cut of the `(&(` and `))` --> attr1=val1)(attr2=val2
-                attrs_conditions = [condition.split("=") for condition in bracketList.split(")(")]
+                attrs_conditions = [tuple(condition.split("=")) for condition in bracketList.split(")(")]
                 return self._multiple_attrs_onelevel_search(base, attrs_conditions)
             else:
                 raise self.PresetReturnRequiredError('search_s("%s", %d, "%s", "%s", %d)' %
```

mateuszmandera replied:
Yeah that should do it

mateuszmandera commented on fakeldap.py:
```diff
@@ -440,22 +449,29 @@ def _add_s(self, dn, record):
 
-                search_attr = attrs.get(search_attr_name)
-                if search_attr == search_attr_value:
+                match = True
+                for search_attr_name, search_attr_value in attrs_conditions:
+                    search_attr = attrs.get(search_attr_name)
+                    this_loop_match = False
+                    if search_attr == search_attr_value:
+                        this_loop_match = True
+                    elif isinstance(search_attr, list):  # if attr is in the format "attr_name": ["value", ]
+                        if len(search_attr) == 1 and search_attr[0] == search_attr_value:
+                            this_loop_match = True
+                    if not this_loop_match:
+                        match = False
+                        break
```
I think we could write this a bit more simply by just doing like this?
```
                match = True
                for search_attr_name, search_attr_value in attrs_conditions:
                    search_attr = attrs.get(search_attr_name)
                    if search_attr == search_attr_value:
                        continue
                    elif isinstance(search_attr, list):  # if attr is in the format "attr_name": ["value", ]
                        if len(search_attr) == 1 and search_attr[0] == search_attr_value:
                            continue
                    else:
                        match = False
                        break
```

ClundXIII replied:
This could be done with minor modification to your code:
```python
                match = True
                for search_attr_name, search_attr_value in attrs_conditions:
                    search_attr = attrs.get(search_attr_name)
                    if search_attr == search_attr_value:
                        continue
                    elif isinstance(search_attr, list):  # if attr is in the format "attr_name": ["value", ]
                        if len(search_attr) == 1 and search_attr[0] == search_attr_value:
                            continue
                        else:
                            match = False
                            break
                    else:
                        match = False
                        break

                if match:
                    result.append((dn, attrs))
```

not sure if that is better

ClundXIII replied:
diff to current:

```diff
diff --git a/fakeldap.py b/fakeldap.py
index 03e8437..1d39aac 100644
--- a/fakeldap.py
+++ b/fakeldap.py
@@ -460,13 +460,15 @@ class MockLDAP(object):
                 match = True
                 for search_attr_name, search_attr_value in attrs_conditions:
                     search_attr = attrs.get(search_attr_name)
-                    this_loop_match = False
                     if search_attr == search_attr_value:
-                        this_loop_match = True
+                        continue
                     elif isinstance(search_attr, list):  # if attr is in the format "attr_name": ["value", ]
                         if len(search_attr) == 1 and search_attr[0] == search_attr_value:
-                            this_loop_match = True
-                    if not this_loop_match:
+                            continue
+                        else:
+                            match = False
+                            break
+                    else:
                         match = False
                         break
```

ClundXIII replied:
also it would probably be better to iterate over `search_attr` :upside_down_face:

ClundXIII replied:
ah, I got it and added option to match list at the same time:

308ecb10570d15d41c28d0bddfdfd0681e437dac

```python

                match = True
                for search_attr_name, search_attr_value in attrs_conditions:
                    search_attr = attrs.get(search_attr_name)
                    if search_attr == search_attr_value:
                        continue
                    elif isinstance(search_attr, list) and search_attr_value in search_attr:  # if attr is in the format "attr_name": ["value", ]
                        continue
                    else:
                        match = False
                        break

                if match:
                    result.append((dn, attrs))
```

mateuszmandera replied:
Ah, yup, that looks good, thanks for fixing the quick snippet i made!

mateuszmandera commented on fakeldap.py:
```diff
@@ -440,22 +449,29 @@ def _add_s(self, dn, record):
-    def _simple_onelevel_search(self, base, filterstr):
-        search_attr_name, search_attr_value = filterstr[1:-1].split('=')
-
+    def _multiple_attrs_onelevel_search(self, base, attrs_conditions):
```
Would be helpful to add a docstring here that `attrs_conditions` are considered to be `AND`-joined because that's the only operation we support

ClundXIII replied:
see 36297324ff0bd2128646333525260cde568d2b16

diff:
```diffdiff --git a/fakeldap.py b/fakeldap.py
index 03e8437..c87d49c 100644
--- a/fakeldap.py
+++ b/fakeldap.py
@@ -450,6 +450,12 @@ class MockLDAP(object):
             return (105,[], len(self.calls), [])
 
     def _multiple_attrs_onelevel_search(self, base, attrs_conditions):
+        """Search the base with tuples of conditions that are "anded" together
+        (all conditions have to be met)
+
+        attrs_conditions -- List of tuples, first element of each tuple is the attribute
+                            name. Second element is the required value.
+        """
         result = []
         for dn, attrs in self.directory.items():
             if dn.endswith(',{}'.format(base)):
```

ClundXIII replied:
ah, I got it and added option to match list at the same time:

308ecb10570d15d41c28d0bddfdfd0681e437dac

```python

                match = True
                for search_attr_name, search_attr_value in attrs_conditions:
                    search_attr = attrs.get(search_attr_name)
                    if search_attr == search_attr_value:
                        continue
                    elif isinstance(search_attr, list) and search_attr_value in search_attr:  # if attr is in the format "attr_name": ["value", ]
                        continue
                    else:
                        match = False
                        break

                if match:
                    result.append((dn, attrs))
```

mateuszmandera replied:
Sounds good to me

mateuszmandera reviewed (commented):
Thanks for adding this functionality! Posted a couple of comments on the code structure but looks good other than that!

mateuszmandera reviewed (approved):
Looks good to me now! @timabbott Can you review next?

ClundXIII commented:
@timabbott ping :)

mateuszmandera commented:
@ClundXIII He's out on vacation this week, so perhaps next week this can get reviewed :smile:

ClundXIII commented:
any chance this will be reviewed next week?

timabbott commented on fakeldap.py:
```diff
@@ -417,12 +417,21 @@ def _search_s(self, base, scope, filterstr, attrlist, attrsonly):
+            elif r_anded.match(filterstr) is not None:
+                bracketList = filterstr[3:-2] # cut of the `(&(` and `))` --> attr1=val1)(attr2=val2
+                attrs_conditions = [tuple(condition.split("=")) for condition in bracketList.split(")(")]
+                return self._multiple_attrs_onelevel_search(base, attrs_conditions)
```
This block would be much better written with Python named capture groups: https://docs.python.org/3/howto/regex.html#non-capturing-and-named-groups

ClundXIII replied:
It is not trivial to capture multiple groups, so I would rather keep the current version (I will force push a quick fix though since the expressions is a bit buggy).

the `regex` module does have a `.captures()` method to capture multiple occurrences, but the code currently uses re.

Reference:
https://stackoverflow.com/questions/5060659/regexes-how-to-access-multiple-matches-of-a-group
https://stackoverflow.com/questions/4963691/regex-with-multiple-groups

timabbott replied:
OK, probably not worth the effort to rework.

timabbott commented:
Posted one comment on the code; also do we need to update the documentation to advertise this feature?

Otherwise lgtm.

ClundXIII commented:
squashed the old fixups, added a new one. Lets take the string `(&(foo=bar)(foo2=bar2))`

The old version matched it as `foo` equals `bar)(foo2=bar2`

The new version correctly iterates and matches `foo` equals `bar` and `foo2` equals `bar2`

timabbott commented:
Looks good, merged, thanks @ClundXIII!
